linked_list/linked_list.py

```diff
@@ -1,128 +1,193 @@
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        if index < 0:
+            return None
+
+        current_node = self.__sentinel.next
+
+        while index > 0:
+            if current_node.value is None:
+                return None
+            current_node = current_node.next
+            index -= 1
+        return current_node.value
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        return self.__sentinel.prev.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        new_node = Node(value, self.__sentinel, self.__sentinel.prev)
+        self.__sentinel.prev = new_node
+        new_node.prev.next = new_node
+        self.__length += 1
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        current_node = self.__sentinel.next
+        max_value = None
+
+        while current_node.value:
+            max_value = max(
+                max_value, current_node.value) if max_value else current_node.value
+            current_node = current_node.next
+
+        return max_value
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def delete(self, value):
-        pass
+        current_node = self.__sentinel.next
+
+        while current_node.value:
+            if current_node.value == value:
+                current_node.prev.next = current_node.next
+                current_node.next.prev = current_node.prev
+                self.__length -= 1
+                return
+            current_node = current_node.next
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def visit(self):
         helper_list = []
-        current = self.head
+        current = self.sentinel.next
 
-        while current:
+        while current.value:
             helper_list.append(str(current.value))
             current = current.next
-        
+
         print(", ".join(helper_list))
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: I think O(n) because delete method is always called on the first node so no need to traverse the list to remove the node(??)
+    # Space Complexity: O(n)
     def reverse(self):

```

Comment:
👍  The time complexity is right, is there a different way to do this in O(1) space complexity?